### PR TITLE
Added publishEndpoint method in Consensus contract

### LINKED ISSUE #85: Validators should publish the end points in consensus contract
- Add a function `publishEndpoint(bytes32 _metachainId, string _service, string _endpoint)`
`msg.sender` is validator.
- The `_metachainId` should be valid, `core.isValidator(msg.sender)` should be true and `!reputation.isSlashed(validator)` should be true.  (not slashed)
- This can be called only by validators
- Emit an event with  `_metachainId, core, validator, service, endpoint`

Service can be `enode`, `ipfs`, etc.

### PR DESCRIPTION
**Pull request by gulshanvasnani**
PR contains changes to add:
1. **publishEndpoint** method.
2. Event decoder.
3. Unit test cases

fixes #85

### CODE REVIEW

File: contracts/consensus/Consensus.sol [hunk 1]
```diff
@@ -661,6 +671,51 @@ contract Consensus is MasterCopyNonUpgradable, CoreLifetimeEnum, MosaicVersion,
...
+        bytes32 _metachainId,
+        string calldata _service,
+        string calldata _endpoint
+    )
+        external
+    {
+        address core = assignments[_metachainId];
+        require(
+            core != address(0),
+            "Invalid metachain id"
```
Comment (0xsarvesh) [inline contracts/consensus/Consensus.sol]:
```suggestion
            "No core exists for the metachain id."
```

Comment (gulshanvasnani) [reply contracts/consensus/Consensus.sol]: Done

File: contracts/consensus/Consensus.sol [hunk 2]
```diff
@@ -34,6 +34,16 @@ contract Consensus is MasterCopyNonUpgradable, CoreLifetimeEnum, MosaicVersion,
 
     using SafeMath for uint256;
 
+    /* Events */
```
Comment (deepesh-kn) [inline contracts/consensus/Consensus.sol:38]: As per the guidelines, add a new line before events.

File: contracts/consensus/Consensus.sol [hunk 3]
```diff
@@ -34,6 +34,16 @@ contract Consensus is MasterCopyNonUpgradable, CoreLifetimeEnum, MosaicVersion,
...
+
+    event EndpointPublished(
+        bytes32 _metachainId,
+        address _core,
+        address _validator,
+        string _service,
+        string _endpoint
```
Comment (deepesh-kn) [inline contracts/consensus/Consensus.sol]: Remove the `_`'s

File: contracts/test/core/SpyCore.sol [hunk 4]
```diff
@@ -123,4 +123,15 @@ contract SpyCore is MasterCopyNonUpgradable, CoreI{
...
+
+    // Note: Implementation will be done once it is required. As the method is
+    // in ICore interface, we would need it, otherwise SpyCore deployment cannot
+    // be done.
+    function isValidator(address _account)
+        external
+        view
+        returns (bool)
+    {
+
```
Comment (deepesh-kn) [inline contracts/test/core/SpyCore.sol]: As its not used in tests, please add an assert to break if its called.
```
// This is not used in test so break
        require(false, "This should not be called for unit tests.");
```

File: contracts/test/core/MockCore.sol [hunk 5]
```diff
@@ -106,4 +106,20 @@ contract MockCore is Core {
     {
         return proposals[_kernelHeight][SENTINEL_PROPOSALS] != bytes32(0);
     }
+
+    function activateValidator(
```
Comment (deepesh-kn) [inline contracts/test/core/MockCore.sol]:
```suggestion
    function addValidator(
```

**Review by deepesh-kn — CHANGES_REQUESTED** [review]
👍 
few small changes.

File: contracts/consensus/Consensus.sol (hunk 2 above)
Comment (gulshanvasnani) [reply contracts/consensus/Consensus.sol:38]: Done.

File: contracts/consensus/Consensus.sol (hunk 3 above)
Comment (gulshanvasnani) [reply contracts/consensus/Consensus.sol]: Done.

File: contracts/test/core/SpyCore.sol (hunk 4 above)
Comment (gulshanvasnani) [reply contracts/test/core/SpyCore.sol]: Done

File: contracts/test/core/MockCore.sol (hunk 5 above)
Comment (gulshanvasnani) [reply contracts/test/core/MockCore.sol]: Done

File: contracts/test/core/SpyCore.sol (hunk 4 above)
Comment (gulshanvasnani) [reply contracts/test/core/SpyCore.sol]: Done

**Review by deepesh-kn — APPROVED** [review]
LGTM 👍

**Pull request merged**